### DISCUSSION TITLE
Strange behavior in docs when using multiple response models

### DISCUSSION THREAD

**Discussion by jordantshaw** [fastapi/fastapi#12966] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union

from fastapi import FastAPI
from pydantic import BaseModel, RootModel


class Person(BaseModel):
    fname: str
    lname: str

class Company(BaseModel):
    name: str

class EntityResponse1(RootModel[Union[list[Person], list[Company]]]):
    root: Union[list[Person], list[Company]]

class EntityResponse2(RootModel[list[Union[Person, Company]]]):
    root: list[Union[Person, Company]]

app = FastAPI()

@app.get("/test1", response_model=EntityResponse1)
def read_root1():
    return {"Hello": "World"}

@app.get("/test2", response_model=EntityResponse2)
def read_root2():
    return {"Hello": "World"}
```

### Description

What is the correct way to have two different response models, each response being an array of objects. 

In my example, I have two different responses, `Entity1` and `Entity2`. I think that `Entity1`  is the correct way to do this. But if you go to the redoc docs, you can see in the example response it just shows an empty array `[]`. 

If I use `Entity2` instead, the redoc docs don't show an empty array anymore, but now the response doesnt seem correct either.

### Operating System

macOS

### FastAPI Version

0.114.1

### Pydantic Version

2.9.2

### Python Version

Python 3.12.6

**Comment by estebanx64** [comment]
In your example I'm seeing two different types of response, the Entity1 is expecting to respond to Either a Person or Company Model.
Entity2 is a mix of both which is different.

You have two ways to add custom responses in a more specific way.

You can add the responses as shown here in the documentation https://fastapi.tiangolo.com/advanced/additional-responses/

or you can use the Pydantic approach https://docs.pydantic.dev/1.10/usage/schema/#schema-customization

```python
from typing import Union

from fastapi import FastAPI
from pydantic import BaseModel, RootModel


class Person(BaseModel):
    fname: str
    lname: str

class Company(BaseModel):
    name: str

class EntityResponse1(RootModel):
    root: Union[list[Person], list[Company]]

    model_config = {
        "json_schema_extra": {
            "examples": [
                # Example 1: List of Person objects
                [
                    {"fname": "John", "lname": "Doe"},
                    {"fname": "Jane", "lname": "Smith"}
                ],
                # Example 2: List of Company objects
                [
                    {"name": "Acme Inc"},
                    {"name": "TechCorp Ltd"}
                ]
            ]
        }
    }

class EntityResponse2(RootModel):
    root: list[Union[Person, Company]]

    model_config = {
        "json_schema_extra": {
            "examples": [
                [
                    {"fname": "John", "lname": "Doe"},
                    {"name": "Acme Inc"},
                    {"fname": "Jane", "lname": "Smith"}
                ]
            ]
        }
    }

app = FastAPI()

@app.get("/test1", response_model=EntityResponse1)
def read_root1():
    return {"Hello": "World"}

@app.get("/test2", response_model=EntityResponse2)
def read_root2():
    return {"Hello": "World"}
```

If you can provide more details about what you need to return it would be helpful, I hope these docs can help you 🤓

**Reply by jordantshaw** [reply]
@estebanx64 thanks for the reply. I am actually wanting to use the first example `Entity1`. Which is expecting either a Persron or Company model. The problem with doing that, is the auto generated docs do not provide a proper "example". Instead it just shows an empty array. 

I understand I could specify my own example using json_schema_extra, but that would not be ideal. The real world response model is quite large and could have many fields. Also, using json_schema_extra will add the examples to the openapi, which is not my desired outcome. I just want the examples geneated in the docs.

**Reply by estebanx64** [reply, marked as answer]
Hi @jordantshaw 👋 

Thanks for being patient.

Debugging your issue it turns out that it is a bug in redoc and not in the way fastapi makes the openapi schema, you can download the openapi schema and it will show the example correctly and not an empty array in a swagger editor for example. 🤓

Here is a link to a related bug, I encourage you to open an issue in redoc uploading some test and your openapi schema.
https://github.com/Redocly/redoc/issues/1638

The only solution, for now, is to describe the responses manually like the links in the documentation I left above to make redoc display them as you want until they fix the bug.

Or use the regular `/docs` swagger as an alternative as they show your Entities as expected.🙇

**Reply by jordantshaw** [reply]
Thank you for helping with this.
